### PR TITLE
Use Jenkins 2.60.1, JDK 8, and git client 3.0.0

### PR DESCRIPTION
**Pull request by MarkEWaite**
Git plugin 4.0.0.  Key items for the checklist:

* [x] unit tests passing
* [x] interactive tests passing
* [x] code review ( @jenkinsci/code-reviewers or @reviewbybees )
* [ ] 2 week beta test
* [x] git client 3.0.0  released with [PR292](https://github.com/jenkinsci/git-client-plugin/pull/292)

### CODE REVIEW

File: src/main/java/hudson/plugins/git/GitChangeSet.java [hunk 1]
```diff
@@ -432,9 +432,6 @@ private boolean hasHudsonTasksMailer() {
 
     private boolean isCreateAccountBasedOnEmail() {
         Hudson hudson = Hudson.getInstance();
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/git/GitChangeSet.java]: Use `jenkins.model.Jenkins` instead? :)

File: Jenkinsfile [hunk 2]
```diff
@@ -1,11 +1,7 @@
 #!groovy
 
-// Don't test plugin compatibility - exceeds 1 hour timeout
+// Test plugin compatbility to recent Jenkins LTS
```
Comment (oleg-nenashev) [inline Jenkinsfile]: typo in "compatbility"

File: pom.xml [hunk 3]
```diff
@@ -3,58 +3,40 @@
...
   <properties>
-    <jenkins.version>1.625.3</jenkins.version>
-    <java.level>7</java.level>
+    <jenkins.version>2.60.3</jenkins.version>
+    <java.level>8</java.level>
     <no-test-jar>false</no-test-jar>
     <concurrency>1C</concurrency>
-    <findbugs.failOnError>false</findbugs.failOnError>
-    <workflow.version>1.14.2</workflow.version>
+    <findbugs.failOnError>true</findbugs.failOnError>
```
Comment (jglick) [inline pom.xml]: Default; just delete.

File: pom.xml [hunk 4]
```diff
@@ -3,58 +3,40 @@
...
+    <maven-project-info-reports-plugin.version>2.9</maven-project-info-reports-plugin.version>
```
Comment (jglick) [inline pom.xml]: Is this needed for something?

File: pom.xml [hunk 5]
```diff
@@ -3,58 +3,40 @@
...
+    <workflow.version>2.4</workflow.version>
```
Comment (jglick) [inline pom.xml]: There is no such thing as a “Workflow version” (or “Pipeline version”) after 1.13. As of 2.x, all plugin versions are independent. Anyway this property seems to be unused, so why not just delete it?

File: pom.xml [hunk 6]
```diff
@@ -127,7 +81,7 @@
     <dependency>
       <groupId>org.jenkins-ci</groupId>
       <artifactId>annotation-indexer</artifactId>
-      <version>1.11</version>
+      <version>1.12</version>
```
Comment (jglick) [inline pom.xml]: Should really be picked up from `jenkins-core` instead. I guess it being an `optional` dependency is a problem. Cue [JENKINS-48578](https://issues.jenkins-ci.org/browse/JENKINS-48578).

File: src/main/java/hudson/plugins/git/util/AncestryBuildChooser.java [hunk 7]
```diff
@@ -55,50 +55,48 @@ public String getAncestorCommitSha1() {
...
 
-                            if (ancestryFilter.isEnabled() && !ancestryFilter.apply(currentRev)) {
-                                continue;
-                            }
+                        if (ageFilter.isEnabled() && !ageFilter.apply(currentRev)) {
+                            continue;
+                        }
 
-                            filteredCandidates.add(currentRevision);
+                        if (ancestryFilter.isEnabled() && !ancestryFilter.apply(currentRev)) {
```
Comment (jglick) [inline src/main/java/hudson/plugins/git/util/AncestryBuildChooser.java]: BTW [whitespace-reduced diff](https://github.com/jenkinsci/git-plugin/pull/565/files?w=1)

File: src/main/java/jenkins/plugins/git/traits/GitBrowserSCMSourceTrait.java [hunk 8]
```diff
@@ -110,8 +110,11 @@ public String getDisplayName() {
          */
         @Restricted(NoExternalUse.class) // stapler
         public List<Descriptor<RepositoryBrowser<?>>> getBrowserDescriptors() {
-            return ((GitSCM.DescriptorImpl) Jenkins.getActiveInstance().getDescriptor(GitSCM.class))
-                    .getBrowserDescriptors();
+            GitSCM.DescriptorImpl descriptor = (GitSCM.DescriptorImpl) Jenkins.getActiveInstance().getDescriptor(GitSCM.class);
+            if (descriptor == null) {
+                return java.util.Collections.emptyList(); // Should be unreachable
```
Comment (jglick) [inline src/main/java/jenkins/plugins/git/traits/GitBrowserSCMSourceTrait.java]: BTW can use `getDescriptorByType`, and as of 2.87 can use `ExtensionList.lookupSingleton` to skip the null check.

File: src/test/java/hudson/plugins/git/SCMTriggerTest.java [hunk 9]
```diff
@@ -277,12 +277,9 @@ private String prepareRepo(ZipFile repoZip) throws IOException {
...
-            public Void call() throws Exception
-            {
-                trigger.run();
-                return null;
-            }
+        Callable<Void> callable = () -> {
+            trigger.run();
+            return null;
         };
         return singleThreadExecutor.submit(callable);
```
Comment (jglick) [inline src/test/java/hudson/plugins/git/SCMTriggerTest.java]: or

```java
return singleThreadExecutor.submit(() -> trigger.run());
```

File: src/test/java/hudson/plugins/git/UserRemoteConfigTest.java [hunk 10]
```diff
@@ -51,13 +51,10 @@ public void credentialsDropdown() throws Exception {
...
     private void assertCredentials(@CheckForNull final Item project, @CheckForNull final String currentCredentialsId, @Nonnull String user, @Nonnull String... expectedCredentialsIds) {
         final Set<String> actual = new TreeSet<>(); // for purposes of this test we do not care about order (though StandardListBoxModel does define some)
-        ACL.impersonate(User.get(user).impersonate(), new Runnable() {
-            @Override
-            public void run() {
-                for (ListBoxModel.Option option : r.jenkins.getDescriptorByType(UserRemoteConfig.DescriptorImpl.class).
-                        doFillCredentialsIdItems(project, "http://wherever.jenkins.io/", currentCredentialsId)) {
-                    actual.add(option.value);
-                }
+        ACL.impersonate(User.get(user).impersonate(), () -> {
```
Comment (jglick) [inline src/test/java/hudson/plugins/git/UserRemoteConfigTest.java]: or use `.as`

File: Jenkinsfile (hunk 2 above)
Comment (MarkEWaite) [reply Jenkinsfile]: Done

File: pom.xml (hunk 3 above)
Comment (MarkEWaite) [reply pom.xml]: Done

File: pom.xml (hunk 4 above)
Comment (MarkEWaite) [reply pom.xml]: Without it, the `mvn site` command that I use complains that the version is not specified.

File: pom.xml (hunk 6 above)
Comment (MarkEWaite) [reply pom.xml]: Without that dependency, I receive a require upper bounds dependencies error:

```
[WARNING] Rule 6: org.apache.maven.plugins.enforcer.RequireUpperBoundDeps failed with message:
Failed while enforcing RequireUpperBoundDeps. The error(s) are [
Require upper bound dependencies error for org.jenkins-ci:annotation-indexer:1.4 paths to dependency are:
+-org.jenkins-ci.plugins:git:4.0.0-SNAPSHOT
  +-com.infradna.tool:bridge-method-annotation:1.17
    +-org.jenkins-ci:annotation-indexer:1.4
and
+-org.jenkins-ci.plugins:git:4.0.0-SNAPSHOT
  +-org.jenkins-ci.main:jenkins-core:2.60.3
    +-org.jenkins-ci:annotation-indexer:1.11
and
+-org.jenkins-ci.plugins:git:4.0.0-SNAPSHOT
  +-org.jenkins-ci.main:jenkins-core:2.60.3
    +-org.jenkins-ci:bytecode-compatibility-transformer:1.8
      +-org.jenkins-ci:annotation-indexer:1.4
and
+-org.jenkins-ci.plugins:git:4.0.0-SNAPSHOT
  +-org.jenkins-ci.main:jenkins-core:2.60.3
    +-org.kohsuke:access-modifier-annotation:1.12
      +-org.jenkins-ci:annotation-indexer:1.4
]
```

I assume that I'm "stuck" with that for now, and can revisit it when JENKINS-48578 provides a BOM for jenkins core.

File: src/main/java/hudson/plugins/git/util/AncestryBuildChooser.java (hunk 7 above)
Comment (MarkEWaite) [reply src/main/java/hudson/plugins/git/util/AncestryBuildChooser.java]: I assume this does not require a change.  The new indentation is correct (as was the old).  The whitespace reduced diff is a nice way to see that the key change was the switch to use a lambda.

File: src/test/java/hudson/plugins/git/SCMTriggerTest.java (hunk 9 above)
Comment (MarkEWaite) [reply src/test/java/hudson/plugins/git/SCMTriggerTest.java]: I'm not yet skilled enough with the Java 8 changes to adapt to this message:

> [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.6.1:testCompile (default-testCompile) on project git: Compilation failure
> [ERROR] /home/mwaite/git/jenkins/git-plugin/src/test/java/hudson/plugins/git/SCMTriggerTest.java:[280,43] incompatible types: java.util.concurrent.Future<capture#1 of ?> cannot be converted to java.util.concurrent.Future<java.lang.Void>

Any hints?

File: src/main/java/hudson/plugins/git/GitSCM.java [hunk 11]
```diff
@@ -1072,7 +1072,7 @@ public EnvVars getEnvironment() {
         buildData.saveBuild(revToBuild);
 
         if (buildData.getBuildsByBranchName().size() >= 100) {
-            log.println("JENKINS-19022: warning: possible memory leak due to Git plugin usage; see: https://wiki.jenkins-ci.org/display/JENKINS/Remove+Git+Plugin+BuildsByBranch+BuildData");
+            log.println("JENKINS-19022: warning: possible memory leak due to Git plugin usage; see: https://wiki.jenkins.io/display/JENKINS/Remove+Git+Plugin+BuildsByBranch+BuildData");
```
Comment (daniel-beck) [inline src/main/java/hudson/plugins/git/GitSCM.java]: Perhaps better as a jenkins.io redirect for link longevity.

Comment (MarkEWaite) [reply src/main/java/hudson/plugins/git/GitSCM.java]: Done in 529928e1e4388a183b461933d8e4f6bfe6ec0b0f .

**Comment by MarkEWaite** [comment]
This pull request needs to wait for its merge until the JENKINS-48061 changes from #577 are merged and released.

**Pull request merged**